core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hadoop;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.BaseMetastoreCatalog;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+
+/**
+ * The HadoopCatalog provides a way to manage the path-based table just like catalog. It uses
+ * a specified directory under a specified filesystem as the warehouse directory, and organizes
+ * two levels directories that mapped to the database and the table respectively. The HadoopCatalog
+ * takes a URI as its root directory, and creates a child directory "iceberg/warehouse" as the
+ * warehouse directory. When creating a table such as $db.$tbl, it creates $db.db/$tbl under
+ * iceberg/warehouse, and put the table metadata into the directory.
+ *
+ * The HadoopCatalog now supports createtable, droptable, the renametable is not supported yet.
+ *
+ * Note: The HadoopCatalog requires the rename operation of the filesystem is an atomic operation.
+ */
+public class HadoopCatalog extends BaseMetastoreCatalog implements Closeable {
+  private static final String ICEBERG_HADOOP_WAREHOUSE_BASE = "iceberg/warehouse";
+  private final Configuration conf;
+  private String warehouseUri;
+
+  public HadoopCatalog(Configuration conf, String warehouseUri) {
+    this.conf = conf;
+
+    if (warehouseUri != null) {
```

Comment:
I don't think that the user should pass `null` to get a default warehouse location. Instead, I think that this class should have a second constructor, `public HadoopCatalog(Configuration conf) { ... }` that creates the default warehouse location and passes it to this constructor. Then this constructor should validate that the warehouse path is non-null using `Preconditions`.